### PR TITLE
Fix parse MySQL binlog event for MaterializeMySQL

### LINKED ISSUE #14781: MaterializeMySQL Data inconsistency between MySQL and Clickhouse
**Describe the bug**
mysql:5.7.31
clickhouse:20.10.1.4635 
I  found data inconsistency between MySQL and Clickhouse when I execute update in MySQL with one transaction with 439129 rows
**How to reproduce**
step1:
    I check that the amount of data is consistent between MySQL and Clickhouse
  mysql> select count(1) from mobile_belong 
    -> ;
+----------+
| count(1) |
+----------+
|   439129 |
+----------+
test-1-118.raipeng.com :) select count(1) from mobile_belong;
SELECT count(1)
FROM mobile_belong
┌─count(1)─┐
│   439129 │
└──────────┘
1 rows in set. Elapsed: 0.043 sec. Processed 439.13 thousand rows, 10.98 MB (10.30 million rows/s., 257.46 MB/s.) 
step2: 
   I executed update in MySQL as below
 mysql> update mobile_belong set province='北京',city='北京';
Query OK, 419371 rows affected (1.98 sec)
Rows matched: 439129  Changed: 419371  Warnings: 0
step3:
 checking data in MySQL and ClickHouse, and found  mysql updated data of 439129 rows successfully ,but,clickhouse only have 438459 rows be updated successfully,remian  439129-438459  not be updated.
mysql> select count(1) from mobile_belong where province='北京' and city='北京';
+----------+
| count(1) |
+----------+
|   439129 |
+----------+
1 row in set (0.16 sec)
test-1-118.raipeng.com :) select count(1) from mobile_belong where province='北京' and city='北京';
SELECT count(1)
FROM mobile_belong
WHERE (province = '北京') AND (city = '北京')
┌─count(1)─┐
│   438459 │
└──────────┘
1 rows in set. Elapsed: 0.068 sec. Processed 523.68 thousand rows, 29.88 MB (7.66 million rows/s., 436.86 MB/s.)

### PR DESCRIPTION
**Pull request by zhang2014**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed bug in parsing MySQL binlog events, which causes `Attempt to read after eof` and `Packet payload is not fully read` in `MaterializeMySQL` database engine

#4006 
fixes #14781 and #14786 and #14251 and #14795 and #14835 and #14829
Fix parse MySQL binlog event for MaterializeMySQL

CC: @tavplubix

### CODE REVIEW

**Comment by BohuTANG** [comment]
Good, passed the test https://gist.github.com/BohuTANG/55cac003a358eb70adce5923be0ee60d

File: src/Core/MySQL/MySQLReplication.cpp [hunk 1]
```diff
@@ -852,68 +853,74 @@ namespace MySQLReplication
...
+            {
                 event = std::make_shared<DryRunEvent>();
-                event->parseHeader(payload);
-                event->parseEvent(payload);
+                event->parseHeader(event_payload);
+                event->parseEvent(event_payload);
                 break;
             }
         }
-        payload.ignoreAll();
```
Comment (tavplubix) [inline src/Core/MySQL/MySQLReplication.cpp:916]: Why `payload.ignoreAll();` or `payload.ignore(CHECKSUM_CRC32_SIGNATURE_LENGTH);` doesn't work and we need `MySQLBinlogEventReadBuffer` to skip the last 4 bytes in packet?

Comment (zhang2014) [reply src/Core/MySQL/MySQLReplication.cpp:916]: We need to extract this logic to accommodate the different checksum algorithms(https://dev.mysql.com/doc/refman/8.0/en/replication-options-binary-log.html#option_mysqld_binlog-checksum) and make it easier to use some read methods. for example: `readStringUntilEOF` and other readUntilEOF.

As a bonus, we don't need to care about checksum in event parsing, and the code is probably more readable.

File: src/Core/MySQL/MySQLReplication.cpp [hunk 2]
```diff
@@ -49,14 +50,13 @@ namespace MySQLReplication
     {
         payload.readStrict(reinterpret_cast<char *>(&binlog_version), 2);
         assert(binlog_version == EVENT_VERSION_V4);
+        server_version.resize(50);
```
[the remark this answers is not in the document]
Comment (zhang2014) [reply src/Core/MySQL/MySQLReplication.cpp:54]: Yes, it needs more unit tests(it should cover all supported events) and I will add it in another PR.

[the remark this answers is not in the document]
Comment (zhang2014) [reply src/Core/MySQL/MySQLReplication.cpp:54]: I added it to the todo list https://github.com/ClickHouse/ClickHouse/issues/4006#issuecomment-657486437(Add more unit tests for MySQL Event #14852 (comment))

File: src/IO/MySQLBinlogEventReadBuffer.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+MySQLBinlogEventReadBuffer::MySQLBinlogEventReadBuffer(ReadBuffer & in_)
+    : ReadBuffer(nullptr, 0, 0), in(in_)
+{
+    nextIfAtEnd();
+}
+
+bool MySQLBinlogEventReadBuffer::nextImpl()
+{
+    if (hasPendingData())
```
[the remark this answers is not in the document]
Comment (zhang2014) [reply src/IO/MySQLBinlogEventReadBuffer.cpp:15]: I checked the code again, it should not be called.

**Review by tavplubix — APPROVED** [review]

**Comment by liangli1213** [comment]
@tavplubix , I saw you have approved the change, and when it will merge to master branch?

**Pull request merged**